core/src/main/java/hudson/Functions.java

```diff
@@ -1079,32 +1082,92 @@ public static String getFooterURL() {
+
+            if (predicate.test(d)) {
+                r.add(new Tag(c.ordinal(), d));
+            }
+        }
+        Collections.sort(r);
+
+        List<Descriptor> answer = new ArrayList<>(r.size());
+        for (Tag d : r) answer.add(d.d);
+
+        return DescriptorVisibilityFilter.apply(Jenkins.get(),answer);
+    }
+
+    /**
+     * Like {@link #getSortedDescriptorsForGlobalConfigByDescriptor(Predicate)} but with a constant truth predicate, to include all descriptors.
+     */
+    public static Collection<Descriptor> getSortedDescriptorsForGlobalConfigByDescriptor() {
+        return getSortedDescriptorsForGlobalConfigByDescriptor(descriptor -> true);
     }
 
     /**
      * @deprecated This is rather meaningless.
      */
     @Deprecated
     public static Collection<Descriptor> getSortedDescriptorsForGlobalConfigNoSecurity() {
-        return getSortedDescriptorsForGlobalConfig(Predicates.not(GlobalSecurityConfiguration.FILTER));
+        return getSortedDescriptorsForGlobalConfigByDescriptor(d -> d != GlobalSecurityConfiguration.FILTER);
     }
 
     /**
-     * Like {@link #getSortedDescriptorsForGlobalConfig(Predicate)} but for unclassified descriptors only.
+     * Descriptors in the global configuration form that users with {@link Jenkins#MANAGE} permission can configure.
+     *
      * @since 1.506
      */
     public static Collection<Descriptor> getSortedDescriptorsForGlobalConfigUnclassified() {
-        return getSortedDescriptorsForGlobalConfig(new Predicate<GlobalConfigurationCategory>() {
-            public boolean apply(GlobalConfigurationCategory cat) {
-                return cat instanceof GlobalConfigurationCategory.Unclassified;
-            }
-        });
+        return getSortedDescriptorsForGlobalConfigByDescriptor(d -> d.getCategory() instanceof GlobalConfigurationCategory.Unclassified && Jenkins.get().hasPermission(d.getRequiredGlobalConfigPagePermission()));
     }
-    
+
+    /**
+     * Descriptors shown in the global configuration form to users with {@link Jenkins#SYSTEM_READ} permission.
+     *
+     * @since TODO
+     */
+    @Restricted(NoExternalUse.class)
+    public static Collection<Descriptor> getSortedDescriptorsForGlobalConfigUnclassifiedReadable() {
+        return getSortedDescriptorsForGlobalConfigByDescriptor(d -> d.getCategory() instanceof GlobalConfigurationCategory.Unclassified && (
+                Jenkins.get().hasPermission(d.getRequiredGlobalConfigPagePermission()) || Jenkins.get().hasPermission(Jenkins.SYSTEM_READ)));
+    }
+
+    /**
+     * Checks if the current security principal has one of the supplied permissions.
+     *
+     * @throws AccessDeniedException
+     *      if the user doesn't have the permission.
+     */
+    @Restricted(NoExternalUse.class)
+    public static void checkAnyPermission(AccessControlled ac, Permission[] permissions) {
+        if (permissions == null || permissions.length == 0) {
```

Comment:
this has different behaviour that `ACL.checkAnyPermission` in the face of an zero length `Permission[]`.  not a bug in itself but would not be obvious and could potentially cause some fun security issues?

Reply:
I think it can be probably removed, it was added to make jelly more forgiving, similar functions in this class have it I believe, I’ll try remove it later on and check it behaves fine

Reply:
this is needed as both checkPermission and checkAnyPermission are always called in l:layout,

Without this check logging in causes:
```
Feb 20, 2020 8:21:21 PM hudson.ExpressionFactory2$JexlExpression evaluate
WARNING: Caught exception evaluating: h.checkAnyPermission(it, permissions) in /jenkins/. Reason: java.lang.NullPointerException
java.lang.NullPointerException
	at hudson.security.ACL.checkAnyPermission(ACL.java:94)
	at hudson.security.AccessControlled.checkAnyPermission(AccessControlled.java:55)
	at hudson.Functions.checkAnyPermission(Functions.java:1164)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
```

note `checkPermission` has the same behaviour